src/Database/LSMTree/Internal/IndexOrdinaryAcc.hs

```diff
@@ -0,0 +1,100 @@
+{- HLINT ignore "Avoid restricted alias" -}
+
+{-|
+    Incremental construction functionality for the general-purpose fence pointer
+    index.
+-}
+module Database.LSMTree.Internal.IndexOrdinaryAcc
+(
+    IndexOrdinaryAcc,
+    new,
+    append,
+    end
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.List (genericReplicate)
+import           Data.Primitive.ByteArray (byteArrayFromListN)
+import           Data.STRef.Strict (STRef, modifySTRef, newSTRef, readSTRef)
+import           Data.Vector (fromList)
+import qualified Data.Vector.Primitive as Primitive (Vector, length)
+import           Data.Word (Word16, Word8)
+import           Database.LSMTree.Internal.Chunk (Baler, Chunk, createBaler,
+                     feedBaler, readBalerRemnant)
+import           Database.LSMTree.Internal.IndexCompactAcc
+                     (Append (AppendMultiPage, AppendSinglePage))
+import           Database.LSMTree.Internal.IndexOrdinary
+                     (IndexOrdinary (IndexOrdinary))
+import           Database.LSMTree.Internal.Serialise
+                     (SerialisedKey (SerialisedKey'))
+import           Database.LSMTree.Internal.Vector (mkPrimVector)
+
+{-|
+    A general-purpose fence pointer index under incremental construction.
+
+    A value @IndexOrdinaryAcc lastKeysRevRef baler@ denotes a partially
+    constructed index with the following properties:
+
+      * The keys pointed to by @lastKeysRevRef@ are the keys that the index
+        assigns to pages, but in reverse order.
+      * The @baler@ object is used by the index for incremental output of the
+        serialised key list.
+-}
+data IndexOrdinaryAcc s = IndexOrdinaryAcc !(STRef s [SerialisedKey]) !(Baler s)
```

Comment:
Lists are probably not the right accumulator type. It would be better if we could use mutable vectors or something like that. See my comments on #296

Reply:
Changed in f84d4de5af557746cf3b9471acc790f1dd6e7218.